core/storage/collection/gae_models.py

```diff
@@ -352,6 +371,24 @@ def _trusted_commit(
                     self.status == constants.ACTIVITY_STATUS_PRIVATE)
             ).put_async()
 
+        snapshot_metadata_model = self.SNAPSHOT_METADATA_CLASS.get(
+            self.get_snapshot_id(self.id, self.version))
+        mentioned_user_ids = (
+            set(self.owner_ids) |
+            set(self.editor_ids) |
+            set(self.voice_artist_ids) |
+            set(self.viewer_ids)
+        )
+        if (
+                len(commit_cmds) == 1 and
+                commit_cmds[0]['cmd'] == feconf.CMD_CHANGE_ROLE
+        ):
+            mentioned_user_ids.add(
```

Comment:
I worry a bit about this -- we are going to forget to do this in the future if a new cmd is introduced or a new field is introduced, right? How can we make it clear to devs and reviewers that if they add a new cmd they need to edit this part of the code as well? Should it be somehow defined declaratively in a place that's closer to the list of cmds so that it is "in view"?

Reply:
Note: not saying the code needs to be changed so that everything will "automatically work" (although that is ideal), but the introduction of a new cmd with a user ID field needs to at least raise enough of a flag for devs/reviewers. Otherwise it's easily missed at review time (there may be an assumption that "this is already taken care of by the general infrastructure").

Reply:
I think I lean away from the first, because it's going to entail updating all the historical cmds and that can get a bit messy.

So, probably the second one is better! (Would you agree?)

Reply:
I agree. @jameesjohn Would Oppiabot be able to handle these kinds of changes? i.e. There is an change to the code of classes that inherit from `BaseChange` class.

Reply:
I have one other thought: 

As I mentioned before we have the change domain objects (all inherit from `BaseChange`) with `ALLOWED_COMMANDS` (constant dictionary that is not saved in the storage), one element of `ALLOWED_COMMANDS` contains `name` (`cmd` in the storage `commit_cmds`), `required_attribute_names`, `optional_attribute_names` and optionally `allowed_values`. I would add another optional value named `user_id_attribute_names` that would contain names of the attributes that contain a user ID. 

Then I would either move these allowed values into the `feconf` so that I can access them from here OR I would import the change domain objects in here (we never import domain objects on the storage level).

Then here I would get all the `user_id_attribute_names` from the `commit_cmd` and save them into the `mentioned_user_ids`.

Reply:
Instead of adding another optional value, why not add a mandatory field to each "ALLOWED_COMMAND" with "user_id_attribute_names"? (It should be a sibling to required_attribute_names.)

Do you think that would work?

For importing the domain objects, perhaps we should commit fully to the idea of "domain objects are independent things, they don't import anything, they can be imported from anywhere" and start thinking about the desired eventual structure and how to move there in stages. Then this could be the first stage.

Reply:
> Instead of adding another optional value, why not add a mandatory field to each "ALLOWED_COMMAND" with "user_id_attribute_names"? (It should be a sibling to required_attribute_names.)
> 
> Do you think that would work?

Hmm, you mean to move the user ID fields from `required_attribute_names` to `user_id_attribute_names`, what if (in the future) some field is in `optional_attribute_names` and also it is a user ID? In this way, I think that the approach with `user_id_attribute_names` as an optional field specifying which fields from  `required_attribute_names` and `optional_attribute_names` are user ID fields is a bit better. But I'm fine with both of these options, just the issue with optional user ID field needs to be somehow resolved.

> For importing the domain objects, perhaps we should commit fully to the idea of "domain objects are independent things, they don't import anything, they can be imported from anywhere" and start thinking about the desired eventual structure and how to move there in stages. Then this could be the first stage.

Yep, okay

Reply:
Oh no, sorry, that's not what I meant. I meant, have user_id_attribute_names contain names of fields that have user_ids. All entries in that list are duplicates of either the required/optional list.

Also we should try to just hash this out, I think. Can you write a brief doc with the different possible formats (and examples of each, plus pros/cons)? Then ping me, and let's look at it for 5 mins and figure out the best option (possibly via a short chat). At the moment I am not entirely certain whether I am fully getting what you are proposing, or vice versa.

> For importing the domain objects, perhaps we should commit fully to the idea of "domain objects are independent things, they don't import anything, they can be imported from anywhere" and start thinking about the desired eventual structure and how to move there in stages. Then this could be the first stage.

This part may need a doc too.

Reply:
> Oh no, sorry, that's not what I meant. I meant, have user_id_attribute_names contain names of fields that have user_ids. All entries in that list are duplicates of either the required/optional list.

Ah, this is what I initially meant too, just that in the case when there is no user ID field I proposed that the `user_id_attribute_names` would be left out, but I think that it makes sense to have it as required and where there is no user ID field it would just be `user_id_attribute_names = []`.

> Also we should try to just hash this out, I think. Can you write a brief doc with the different possible formats (and examples of each, plus pros/cons)? Then ping me, and let's look at it for 5 mins and figure out the best option (possibly via a short chat). At the moment I am not entirely certain whether I am fully getting what you are proposing, or vice versa.

I don't have any other proposal except for the aforementioned one. I don't think that there is a need for a doc.

> This part may need a doc too.

Yes, will create it soon.

Reply:
OK, sgtm. I agree with the empty list, thanks for clarifying.

Also, not sure if there are further action items for me at the moment with regards to this PR, so I'm deassigning myself for now. But please feel free to reassign me once you're ready for a full review! (Though we might need to hash out the second part of the above discussion first.)

Reply:
Done.